core/src/main/java/org/apache/druid/indexer/partitions/HashedPartitionsSpec.java

```diff
@@ -160,7 +160,7 @@ public Integer getNumShards()
   @Override
   public String getForceGuaranteedRollupIncompatiblityReason()
   {
-    return getNumShards() == null ? NUM_SHARDS + " must be specified" : FORCE_GUARANTEED_ROLLUP_COMPATIBLE;
+    return FORCE_GUARANTEED_ROLLUP_COMPATIBLE;
```

Comment:
https://druid.apache.org/docs/latest/ingestion/native-batch.html#hash-based-partitioning needs to be updated to say that `numShards` is no longer required

Reply:
Updated the docs to mark `numShards` as optional and added a description of the new cardinality scan phase.